File: LeetCode/C++/47.permutations-ii.cpp
/*
 * @lc app=leetcode id=47 lang=cpp
 *
 * [47] Permutations II
 */

// @lc code=start
/* 
 * Time : 94.65%, 3ms
 * Space : 89.52%, 8.5MB
 */
# include <vector>
# include <algorithm>
# include <iostream>

using namespace std;

class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> ret;
        sort(nums.begin(), nums.end());
        do {
            ret.push_back(vector<int>(nums));
        } while (next_permutation(nums.begin(), nums.end()));
        // for (auto & i : ret){
        //     for (auto & n : i){
        //         cout << n << ' ';
        //     } 
        //     cout << endl;
        // }
        return ret;
    }
};

// int main(void){
//     Solution s;
//     vector<int> v = {1, 1, 2, 3, 4};
//     s.permuteUnique(v);
// }
// @lc code=end

File: LeetCode/Rust/47_permutations-ii.rs
/*
 * @lc app=leetcode id=47 lang=rust
 *
 * [47] Permutations II
 */

// @lc code=start
// impl Solution {
//     pub fn permute_unique(nums: Vec<i32>) -> Vec<Vec<i32>> {
//         nums.sort();
//         println!("{}", nums);
//     }
// }

fn next_permutation(mut nums: & Vec<i32>) -> bool {
    let mut i = nums.len() - 2;
    // while i >= 1 {
    //     if &nums[i - 1] < &nums[i] {

    //     }
    // }
    
    // for i in (0..(nums.len() - 1)).rev() {
    //     print!("{}", &nums[i]);
    // }
    true
}

fn permute_unique(nums: Vec<i32>) /*-> Vec<Vec<i32>>*/ {
    let len = nums.len();
    let ret: Vec<Vec<i32>> = Vec::new();
    let mut tmp = Vec::with_capacity(len);
    if len == 0 || len == 1 {
        ret.append(& mut tmp);
        // return ret;
    }
    tmp.copy_from_slice(&nums[..]);
    tmp.sort();
    next_permutation(& mut tmp);
    // println!("{:?}", tmp);
    
}

fn main(){
    permute_unique(vec![3, 2, 1]);
}

// @lc code=end

